### DISCUSSION TITLE
Simulating limbdarkening effect

### DISCUSSION THREAD

**Discussion by TobyOnGithub** [ImageMagick/ImageMagick#6062] [Help]
### ImageMagick version

7.1. 0-61

### Operating system, version and so on

Linux Ubuntu 22.04

### Description

Question / Problem:

I am a linux user and installed MyPaint and GIMP a few days ago and cannot figure out the following. I would like to draw a white circle on a black background. In full-screen mode the white circle should cover most of the screen. The circle should be homogenous with the white-level uniform across the circle. This is part one and likely easy. Now the tricky part 2: the circle should now have a gradient with a centre-region with white-level as in part 1 and then decreasing linearly towards the limb. Basically, I would like to simulate the so-called “limb-darkening” effect ([Limb darkening - Wikipedia](https://en.wikipedia.org/wiki/Limb_darkening)) as seen for our Sun: bright at the centre and darker towards the limb. The limb-darkening effect can be simulated using different brightness-profile laws for the brightness change (linear, quadratic, square-root, log. etc). If we can start with the linear part and then add complexity that would be cool.

**Comment by snibgo** [comment]
What version of IM, on what platform? I assume IM v7 on Window BAT.

> ... draw a white circle on a black background. In full-screen mode the white circle should cover most of the screen. The circle should be homogenous with the white-level uniform across the circle. 

I suggest:
```
magick ^
  -size 201x201 xc:Black ^
  -fill White ^
  -draw "circle 100,100,100,1" ^
  x0.png
```
![x0](https://user-images.githubusercontent.com/33812368/218720937-312d86db-7afd-4f8d-934e-7e5b9b70f9cd.png)

This is Windows BAT syntax. For bash, change line-continuation caret `^` to backslash `\`.

Note that the width and height are equal, and they are odd numbers. The centre is at (100,100) and the circle passes through (100,1).

> ... the circle should now have a gradient with a centre-region with white-level as in part 1 and then decreasing linearly towards the limb.

I suggest:
```
magick ^
  -size 201x201 radial-gradient:white-black ^
  x1.png
```
![x1](https://user-images.githubusercontent.com/33812368/218721069-61533486-c74a-4b41-b168-bc4888b17657.png)

This is a linear gradient from white in the centre to black at the edges. We can do arithmetic on this. For experiments, I suggest using "-fx". If your IM is old, this will be slow. If your IM is newer than 7.1.0-21 it will be faster.

For example, `u` is the pixel value on a scale of 0 to 1. We raise it to the power `0.2`, which will lighten mid-grays:
```
magick ^
  -size 201x201 radial-gradient:white-black ^
  -fx "pow(u,0.2)" ^
  x2.png
```
`-fx` will be slower than a single "-evaluate":
```
magick ^
  -size 201x201 radial-gradient:white-black ^
  -evaluate pow 0.2 ^
  x2.png
```
![x2](https://user-images.githubusercontent.com/33812368/218721686-53f69f0d-8303-475a-9a19-3c23c6de973e.png)

... but has the advantage of flexibility, which is useful when experimenting.

Does that get you on the path?

**Comment by snibgo** [comment]
Using the formulae from https://en.wikipedia.org/wiki/Limb_darkening :
```
set omega=0.27*pi/180
set A1=-0.47
set A2=-0.23

magick ^
  -size 201x201 radial-gradient:white-black ^
  -fx "st = (1-u)*%omega%; so = sin(%omega%); cv = sqrt (1 - pow(st/so,2) ); 1 + (1-cv)*(%A1% + %A2%*(1-cv))" ^
  x4.png
```

![x4](https://user-images.githubusercontent.com/33812368/218766462-dc407a71-b08b-45a6-990c-90f9b0d41530.png)

Don't trust my mathematics. I may have got it wrong.

**Reply by TobyOnGithub** [reply]
wow!! this looks promising!! I am working on a Ubuntu linux machine. I tried a "man magick" but Imagemagick seems not installed. But this is not true. If I do a "convert -version", I get a return like

tobiash@laptop:~$ convert -version
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(4.5) 
Delegates (built-in): bzlib djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

This is a strong indication for that IM is installed and its version 6.9.11-60.

Are you a linux user?

**Reply by TobyOnGithub** [reply]
Hi Alan, I am now looking in some more detail on your method. There might be an error in the math that you implemented, but that is not relevant at the moment. What is important is that your methodology is promising. Question: the wikipage on limb-darkening depicts theta as an independent parameter that describes where on the circle we are. I guess that the -fx switch allows the definition of a function, but where is the independent parameter in that syntax? I only see constants in your math terms.

**Comment by snibgo** [comment]
I use Windows, sometimes (rarely) bash:
```
omega=0.27*pi/180
A1=-0.47
A2=-0.23

convert \
  -size 201x201 radial-gradient:white-black \
  -fx "st = (1-u)*${omega}; so = sin(${omega}); cv = sqrt (1 - pow(st/so,2) ); 1 + (1-cv)*(${A1} + ${A2}*(1-cv))" \
  x4s.png
```

**Reply by TobyOnGithub** [reply]
thank you -- in the meantime I figured out that IM 6 command line API has the following commands "animate, compare, composite, conjure, convert, display, identify, import, mogrify, montage, stream". The command montage seems to work for your first example. 
Is there any chance to output in vector graphic format SVG instead of PNG?

**Comment by snibgo** [comment]
Montage does something different. I suggest "convert".

ImageMagick is a raster image processor, not a vector processor.

Of course, you could simply write the SVG yourself. I don't know if SVG can do the maths you need.

**Reply by TobyOnGithub** [reply]
yes, convert command is better. and if I increase the size of the image I can get a smooth stellar face on my laptop when using full screen. Thank you!! That was very helpful!! However, I still have to study your code - especially the last one. :-) Certainly very helpful!!

**Comment by snibgo** [comment]
> Question: the wikipage on limb-darkening depicts theta as an independent parameter that describes where on the circle we are.

Quick answer: `u`.

Long answer:

I took the angular diameter of the sun as seen from the earth as 0.54 degrees. The angular radius is half that. `omega=0.27*pi/180` is the angular radius in radians. Note that "`0.27*pi/180`" is a string that will be re-evaluated at every pixel, which is inefficient. And then we re-calculate `sin(%omega%)` at every pixel, which is a bigger performance problem. When the maths is correct, we can improve performance, if needed.

For performance, expressions that are constant for all pixels can be evaluated in nested `%[fx:...]` expressions, like this:
```
magick \
  -size 201x201 radial-gradient:white-black \
  -fx "st = (1-u)*%[fx:${omega}]; so = %[fx:sin(${omega})]; cv = sqrt (1 - pow(st/so,2) ); 1 + (1-cv)*(${A1} + ${A2}*(1-cv))" \
  x5s.png
```

`u` is a value at each pixel, loosely speaking the "whiteness". It is 0.0 at the edge of the sun and 1.0 at the centre. So `(1-u)` is 1.0 at the edge and 0.0 at the centre. At any pixel in the sun, `(1-u)*%omega%` is approximately the angular distance of that pixel from the centre of the sun, and also the sine of that angular distance (accurate to 5 decimal places). Hence `sin(theta)` is ` st = (1-u)*%omega`.

If you want a physical representation of the brightness of the sun on a sRGB computer display, I think you should do two things. First, assume the calculations are in linear RGB and then convert to sRGB. Before writing the output, insert this line:
```
  -set colorspace RGB -colorspace sRGB \
```
The second thing is to check whether your computer display has a correct sRGB transform.

**Reply by TobyOnGithub** [reply]
We had some testing in the lap yesterday. Basically we found the following: we found that my laptop screen has a built-in "limbdarkening" effect. During the transit of a star with no limbdarkening no flat-bottom curve could be achieved. There was always more or less a curvature. We tested this by using a very small flat-field screen for other optical experimentation and voila we got a flat bottom on first trial. This means there must be some dependency of screen intensity to incident angle. Some screens have this effect very strongly to avoid other to look into one own laptop screen from behind at an angle. I guess the same effect is experienced here though less extent. Now, my question is: is it possible to make my laptop screen appear completely flat or homogeneous in brightness?

**Comment by snibgo** [comment]
> Now, my question is: is it possible to make my laptop screen appear completely flat or homogeneous in brightness?

Yes, in principle. Just multiply your displayed image by a compensating image.

Suppose you display an image that is entirely white to your screen. With no other ambient light, when you view the screen from a certain fixed point, the apparent lightness varies, for example the corners are darker than the centre. Grab a camera that is known to be linear. (That means: when the raw image is white-balanced and debayered to a *linear* colorspace, the pixel values are proportional to the light energy received. Probably most good quality modernish DSLRs qualify. See [Linear camera raw](http://im.snibgo.com/lincamraw.htm).)

(EDIT: This process also assumes that your camera, and especially the lens, has no vignetting.)

From the fixed point, take a photo of the screen showing the white image. Make it slightly out of focus. Ensure that the image has not clipped. Develop this to a linear image, eg using `dcraw -w -o 0 -4`. You must work from the raw image, not the out-of-camera JPEG. In our example, the corners of the photo will be darker than the centre. Crop the photo to the appropriate boundaries.

Now, when you want to display an image, first *divide* it by the photo that has been resized to be the same size. The result, when viewed from the same fixed point, will have the appropriate compensation. You can check this by displaying a white image that has been divided, and taking a photo as before. This time, the photo should have a constant brightness.

Ensure your IM has HDRI. Beware that dividing one image by another may result in values that are greater than 100%. If so, divide by some number to bring them within range. If we divide by the maximum value, the result will be exactly 100%:
```
magick inpImg.tiff compensatImg.tiff -compose DivideSrc -composite -evaluate Divide %[fx:Maxima] out.tiff
```
I think (but I'm not sure about this) that the arithmetic should be done in linear colorspace. So, if inpImg.tiff is sRGB, then do this:
```
magick inpImg.tiff -colorspace RGB compensatImg.tiff -compose DivideSrc -composite -evaluate Divide %[fx:Maxima] -colorspace sRGB out.tiff
```
[EDIT 2: The above assumes that the variation of screen intensity by incident angle is independent of the input intensity. This isn't true for my laptop. At maximum brightness, there is virtually no variation by angle, but at 50% intensity I get falloff to 40% in the corners.]

**Comment by TobyOnGithub** [comment]
Thank you -- that sounds like a flatfielding process which I am somewhat
familiar with.

Let's take one step at a time.

You write "From the fixed point, take a photo of the screen showing the
white image. Ensure that the image has not clipped. Develop this to a
linear image, eg using dcraw -w -o 0 -4. You must work from the raw image,
not the out-of-camera JPEG. In our example, the corners of the photo will
be darker than the centre. Crop the photo to the appropriate boundaries."

I might be able to borrow a DSLR camera. But I am not at all tech savvy in
operating those monsters.

So, when I turn on the camera the first thing to figure out is how to set
the camera to store images in RAW format, right? I guess that is an easy
task. I understand that I should not care about the JPEG image as it is
compressed and does not reflect what I measured in reality.

Are there any other settings that I should apply or can I just shoot an
image from the default settings that apply after turning on the camera?

Just turn on the camera, store images in RAW format and then take the image
of my laptop screen?

This brings me to the next issue: You write "ensure that the image has not
clipped". What does that mean? Does that mean that the image in the DSLR
camera should cover all the laptop screen including parts of the laptop
screen plastic frame?

Best, Toby
